### ISSUE TITLE
LogDensityProblems ambiguity for DynamicHMC

### ISSUE THREAD

**Issue by cpfiffer** [TuringLang/Turing.jl#730]
The `LogDensityProblems.logdensity` call seems to be ambiguous. What's the correct fix here?

MWE:

```julia
using DynamicHMC, Turing

chn = sample(gdemo_default, DynamicNUTS(2000));
```

Error:

```julia
ERROR: LoadError: MethodError: LogDensityProblems.logdensity(::Type{LogDensityProblems.ValueGradient}, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Array{Float64,1}) is ambiguous. Candidates:
  logdensity(::Type{LogDensityProblems.ValueGradient}, ℓ::Turing.Inference.FunctionLogDensity, x) in Turing.Inference at /home/cameron/.julia/dev/Turing/src/Turing.jl:135
  logdensity(::Type{LogDensityProblems.ValueGradient}, ℓ, x::AbstractArray{T,1}) where T in LogDensityProblems at /home/cameron/.julia/packages/LogDensityProblems/sIFw5/src/problem.jl:46
Possible fix, define
  logdensity(::Type{LogDensityProblems.ValueGradient}, ::Turing.Inference.FunctionLogDensity, ::AbstractArray{T,1})
Stacktrace:
 [1] phasepoint_in(::DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Diagonal{Float64,Array{Float64,1}},LinearAlgebra.LowerTriangular{Float64,LinearAlgebra.Diagonal{Float64,Array{Float64,1}}}}}, ::Array{Float64,1}, ::Array{Float64,1}) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/hamiltonian.jl:149
 [2] #NUTS_init#18(::Array{Float64,1}, ::GaussianKE{LinearAlgebra.Diagonal{Float64,Array{Float64,1}},LinearAlgebra.LowerTriangular{Float64,LinearAlgebra.Diagonal{Float64,Array{Float64,1}}}}, ::Array{Float64,1}, ::Int64, ::DynamicHMC.InitialStepsizeSearch, ::ReportIO{Base.TTY}, ::Function, ::MersenneTwister, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:139
 [3] NUTS_init(::MersenneTwister, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:138
 [4] #NUTS_init_tune_mcmc#20(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::MersenneTwister, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:284
 [5] NUTS_init_tune_mcmc(::MersenneTwister, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:284
 [6] #NUTS_init_tune_mcmc#21(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:294
 [7] NUTS_init_tune_mcmc(::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/cameron/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:294
 [8] sample(::Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#617#128")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}}, ::DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}) at /home/cameron/.julia/dev/Turing/src/inference/dynamichmc.jl:65
 [9] top-level scope at none:0
```

**Comment by xukai92** [comment]
I also meet this error when I was trying to compare against DynamicHMC. I guess @mohamed82008 might know because he recently reported some performance evaluation against DynamicHMC.

**Comment by yebai** [comment]
It seems there are some internal changes in `DynamicHMC`, which caused this issue. This should be fixed by 066d028.
